homeassistant/components/sensor/tank_utility.py

```diff
@@ -0,0 +1,134 @@
+"""
+Support for the Tank Utility propane monitor.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.tank_utility/
+"""
+
+import datetime
+import logging
+
+import requests
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import (CONF_DEVICES, CONF_EMAIL, CONF_PASSWORD,
+                                 STATE_UNKNOWN)
+from homeassistant.helpers.entity import Entity
+
+
+REQUIREMENTS = [
+    "tank_utility==1.4.0"
+]
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = datetime.timedelta(hours=1)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_EMAIL): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Required(CONF_DEVICES): vol.All(cv.ensure_list, vol.Length(min=1))
+})
+
+SENSOR_TYPE = "tank"
+SENSOR_ROUNDING_PRECISION = 1
+SENSOR_UNIT_OF_MEASUREMENT = "%"
+SENSOR_ATTRS = [
+    "name",
+    "address",
+    "capacity",
+    "fuelType",
+    "orientation",
+    "status",
+    "time",
+    "time_iso"
+]
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Tank Utility sensor."""
+    from tank_utility import auth
+    email = config.get(CONF_EMAIL)
+    password = config.get(CONF_PASSWORD)
+    devices = config.get(CONF_DEVICES)
+    token = auth.get_token(email, password)
```

Comment:
Can the token go invalid?

Reply:
It can, and in fact it's invalidated every 24 hours (silly API).  The library handles [caching the token](https://github.com/krismolendyke/tank-utility/blob/d8919d69996dbd8ff199f4e4c4d3a9386158d1ff/src/tank_utility/auth.py#L79) and this sensor [requests a new token when the current one is invalid](https://github.com/home-assistant/home-assistant/pull/9132/files#diff-7e5cee4f3743b7eb48fbc994ead0aa32R115).

Reply:
How you can detect wrong user/password? I think you should check that here.

Reply:
I can do that here because `get_token` raises an error if a token can't be retrieved.  I handle a similar case [during `update`](https://github.com/home-assistant/home-assistant/pull/9132/files#diff-7e5cee4f3743b7eb48fbc994ead0aa32R114) and could take that approach here, too.  I guess that should log an error?  Anything else that `setup_platform` should do if it's been configured with bad credentials?

Reply:
Yeah log a error and return `False`